src/transformers/models/llama/modeling_llama.py

```diff
@@ -889,8 +906,11 @@ def forward(
             raise ValueError("You have to specify either input_ids or inputs_embeds")
 
         past_key_values_length = 0
-        if past_key_values is not None:
-            past_key_values_length = past_key_values[0][0].shape[2]
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_key_values_length = past_key_values.get_seq_length()
```

Comment:
```suggestion
        if use_cache:
            use_legacy_cache = not isinstance(past_key_values, Cache) and past_key_values is not None
            if use_legacy_cache:
                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
            elif past_key_values is None:
                past_key_values = DynamicCache()
            past_key_values_length = past_key_values.get_seq_length()
```

This has probably been discussed already but here I am: here in the first forward pass when using `generate`, we have `past_key_values` being None, and `use_legacy_cache` evaluates to `True`. This results in `from_legacy_cache` and `to_legacy_cache` being called at each decoding iteration. Is this intended for now?

Alternatively `DynamicCache()` may be initialized in `generation/utils.py` in case we care only about `generate` using the new cache implementation.

it's not clear to me why, but the proposed suggestion I give actually makes generation slower. Any idea what's happening?

Reply:
Hey @fxmarty! Glad to hear that it is compatible with the optimum ONNX export 💛 

Defaulting to the old cache type is intentional: for all intents and purposes, if we were to use (and return) a `DynamicCache` by default it would technically be an API change, so it has to go through a deprecation cycle. However, since caching touches many types of models and is used in lots of places, I want users to experiment with the feature and report bugs before placing the warning (like you just did with optimum ONNX export!). For instance, I'm fairly confident that we need code changes for the cache to support encoder-decoder models -- it has no room for the cross-attention atm. 

As for your suggestion being slower: no clue 👀 The conversion is pretty fast though, no tensors are created, so it should be cheap to leave it as is for a couple of minor `transformers` versions